### PR TITLE
[TP][Inference] Add decompose for matmul op

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110833
* #110751

When we set the bias to False in the linear, it will trigger `matmul` during the decomposition of  `linear`. So we want to add matmul op in the decomposition as well. Also since we didn't cover this case in unit test, we add a new unit test case so that we have it covered in our UT.

cc @wanchaol @XilunWu @tianyu-l

### CODE REVIEW

**Comment by Chillee** [comment]
Just decompose all CompositeImplicitAutograd ops?

https://github.com/pytorch/pytorch/blob/main/torch/fx/experimental/proxy_tensor.py#L276-L281

**Comment by fduwjj** [comment]
@Chillee We need to do some benchmark for Eager Perf before we remove the op check. But I agree, eventually we want "decompose all CompositeImplicitAutograd ops"

**Comment by Chillee** [comment]
@fduwjj Decomposing all CompositeImplicitAutograd ops has no effect on eager perf (in terms of kernels launched).

**Comment by fduwjj** [comment]
@Chillee And no CPU overhead as well?

**Comment by Chillee** [comment]
@fduwjj Well, the alternative is to handle each CompositeImplicitAutograd op individually in DTensor, which seems like wasted effort. You've already submitted 2 PRs to enable CompositeImplicitAutograd ops - even just for transformer inference I suspect you'll need to submit 3 more that Hamid hasn't even run into yet (probably `aten::softmax`, `aten::scaled_dot_product_attention`, etc.)

**Comment by fduwjj** [comment]
@Chillee I am not against what you proposed and we eventually will do that. But I guess right now, we want it to be gated. 

And to your point, acutally, I run a local script which run the transformer model E2E (basically what Hamid is testing and I should have run it last Friday so we will not end up in a second PR) and it works. So I don't think we need to enable for the ops you mentioned (at least for his use case) because we don't use DTensor to handle all ops within the model code, e.g., ops like `aten::scaled_dot_product_attention`.
